### PR TITLE
neural network solver for single DBlocks: perfect foresight, shocks

### PR DESCRIPTION
**Pull request by sbenthall**
A neural network solver in `ann.py` and a demonstration of how it works with the approximate lifetime reward loss function.

The automated test scales the torch network size and training data down a lot because we want the test runtime to be low and GitHub actions doesn't have CUDA (does it?). But the code is designed to cleanly use CUDA when available.

The initial version of this PR deals only with perfect foresight. Some TODOs for this work are:

- [x] better support for automatically generating grids of training and visualization data.
- [x] handling models with exogenous shocks. (this requires a lot of shock values to be passed as 'training data')
- [x] testing on more than one DBlock, preferably with a test that's so simple it can be used to validate that things are working in an automated test.

### CODE REVIEW

File: tests/test_ann.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import skagent.ann as ann
+import skagent.grid as grid
+import skagent.models.perfect_foresight as pfm
+import torch
+
+import unittest
+
+## CUDA handling
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+print(torch.cuda.is_available())
```
[the remark this answers is not in the document]
Comment (alanlujan91) [reply tests/test_ann.py:10]: Using this to test if cuda is installed in system. Currently in my system (Windows) it is not installed, so keeping this until I can figure out how to install cuda and this prints true

**Comment by sbenthall** [comment]
@alanlujan91 I think this PR is ready to merge, though there's a lot of performance issues outstanding. I think they should go in separate PRs. Have you been able to look this over to your satisfaction yet?
